src/Microsoft.NetCore.Analyzers/Core/Security/DoNotReferSelfInSerializableClass.cs

```diff
@@ -0,0 +1,191 @@
+                isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                description: s_Description,
+                helpLinkUri: null,
+                customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    var compilation = compilationStartAnalysisContext.Compilation;
+                    var serializableAttributeTypeSymbol = WellKnownTypes.SerializableAttribute(compilation);
+
+                    if (serializableAttributeTypeSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var nonSerializedAttribute = WellKnownTypes.NonSerializedAttribute(compilation);
+
+                    if (nonSerializedAttribute == null)
+                    {
+                        return;
+                    }
+
+                    var forwardGraph = new ConcurrentDictionary<ISymbol, HashSet<ISymbol>>();
+                    var invertedGraph = new ConcurrentDictionary<ISymbol, HashSet<ISymbol>>();
+
+                    // It keeps the out Degree of every vertex in the invertedGraph, which is corresponding to the in Degree of the vertex in forwardGraph.
+                    var inDegree = new ConcurrentDictionary<ISymbol, int>();
+
+                    // It Keeps the out degree of every vertex in the forwardGraph, which is corresponding to the in Degree of the vertex in invertedGraph.
+                    var outDegree = new ConcurrentDictionary<ISymbol, int>();
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        (SymbolAnalysisContext symbolAnalysisContext) =>
+                        {
+                            var classSymbol = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+
+                            if (!classSymbol.HasAttribute(serializableAttributeTypeSymbol))
+                            {
+                                return;
+                            }
+
+                            var fields = classSymbol.GetMembers().OfType<IFieldSymbol>().Where(s => !s.HasAttribute(nonSerializedAttribute) &&
+                                                                                                    !s.IsStatic &&
+                                                                                                    s.Type.IsInSource());
```

Comment:
> s.Type.IsInSource()

Will this handle a constructed type such as `class A { List<A> field; }`  or we don't flag this case?